rx-java2-gen/src/main/java/io/vertx/reactivex/impl/WriteStreamSubscriber.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.vertx.reactivex.impl;
+
+import io.reactivex.FlowableSubscriber;
+import io.vertx.core.Handler;
+import io.vertx.core.streams.WriteStream;
+import org.reactivestreams.Subscription;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * @author Thomas Segismont
+ */
+public class WriteStreamSubscriber<R, T> implements FlowableSubscriber<R> {
+
+  private final WriteStream<T> writeStream;
+  private final Consumer<Throwable> onError;
+  private final Runnable onComplete;
+  private final Function<R, T> adapter;
+
+  private Subscription subscription;
+  private boolean drainHandlerSet;
+
+  public WriteStreamSubscriber(WriteStream<T> writeStream, Function<R, T> adapter, Consumer<Throwable> onError, Runnable onComplete) {
+    Objects.requireNonNull(writeStream, "writeStream");
+    Objects.requireNonNull(adapter, "adapter");
+    Objects.requireNonNull(onError, "onError");
+    Objects.requireNonNull(onComplete, "onComplete");
+    this.writeStream = writeStream;
+    this.adapter = adapter;
+    this.onError = onError;
+    this.onComplete = onComplete;
+  }
+
+  @Override
+  public void onSubscribe(Subscription subscription) {
+    this.subscription = subscription;
+    subscription.request(256);
+  }
+
+  @Override
+  public void onNext(R r) {
+    writeStream.write(adapter.apply(r));
+    if (writeStream.writeQueueFull()) {
+      Handler<Void> h;
+      synchronized (this) {
+        if (drainHandlerSet) {
+          h = null;
+        } else {
+          drainHandlerSet = true;
+          h = v -> subscription.request(1);
+        }
+      }
+      if (h != null) {
+        writeStream.drainHandler(v -> {
+          synchronized (this) {
```

Comment:
why can't this lambda be part of `h` created in the synchronised block ?

Reply:
To minimize actions in the synchronized block.

Reply:
this is actually a lambda capture and not execute. The lambda will be actually called by the drain handler, so there isn't more less work done and actually is it less performant because it will instantiate two lambdas instead of one.

Reply:
actually this lambda could even been created once because it's not capturing anything from the local scope and only updates the object fields.